### PR TITLE
Refactor Adapters

### PR DESCRIPTION
**Pull request by bojiang**
## Types of changes

- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature and improvements (non-breaking change which adds/improves functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Code Refactoring (internal change which is not user facing)
- [ ] Documentation
- [ ] Test, CI, or build

## Component(s) if applicable
- [ ] BentoService (service definition, dependency management, API input/output adapters)
- [ ] Model Artifact (model serialization, multi-framework support)
- [ ] Model Server (mico-batching, dockerisation, logging, OpenAPI, instruments)
- [ ] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [ ] BentoML CLI

## Checklist:

- [x] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/_version.py [hunk 1]
```diff
@@ -247,10 +247,7 @@ def git_pieces_from_vcs(tag_prefix, root, verbose, run_command=run_command):
         raise NotThisMethod("'git rev-parse' failed")
     full_out = full_out.strip()
 
-    pieces = {}
-    pieces["long"] = full_out
-    pieces["short"] = full_out[:7]  # maybe improved later
-    pieces["error"] = None
+    pieces = {"long": full_out, "short": full_out[:7], "error": None}
```
Comment (parano) [inline bentoml/_version.py:250]: what's the reason for this change?  this is actually not our code, it is from the versioneer project

File: bentoml/adapters/default_output.py [hunk 2]
```diff
@@ -39,12 +44,8 @@ def detect_suitable_adapter(result, slices=None):
     except ImportError:
         pass
 
-    if slices is not None:
-        for s in slices:
-            if s:
-                return detect_suitable_adapter(result[s])
-    if isinstance(result, (list, tuple)):
-        return detect_suitable_adapter(result[0])
+    # if isinstance(result, (list, tuple)):
```
Comment (parano) [inline bentoml/adapters/default_output.py:47]: remove?

File: bentoml/service.py [hunk 3]
```diff
@@ -233,50 +337,38 @@ def api_decorator(
...
-
     """
 
     def decorator(func):
         _api_name = func.__name__ if api_name is None else api_name
         validate_inference_api_name(_api_name)
 
         if input is None:
-            # support bentoml<=0.7
+            # Support passing the desired adapter without instantiation
```
Comment (parano) [inline bentoml/service.py:347]: nit: "Raise error when input adapter class passed without instantiation"

**Comment by bojiang** [comment]
Moved to https://github.com/bentoml/BentoML/pull/1070
